Bluetooth: Mesh: Sensor Server and Client sample (#2513)
nrfconnect/sdk-nrf: samples/bluetooth/mesh/sensor_server/README.rst

.. _bluetooth_mesh_sensor_server:

> Bluetooth: Mesh Sensor Server

Reviewer: ```suggestion
Bluetooth: Mesh sensor server
```

###############################

> The Bluetooth Mesh :ref:`bt_mesh_sensor_srv_readme` sample demonstrates how to set up a basic :ref:`bt_mesh_sensor_srv_readme` model application that provides sensor data to one :ref:`bt_mesh_sensor_cli_readme` model.

Reviewer: ```suggestion
The Bluetooth Mesh sensor server sample demonstrates how to set up a basic Mesh sensor server model application that provides sensor data to one :ref:`Mesh sensor client <bt_mesh_sensor_cli_readme>` model.
```

> In this sample the on-chip ``TEMP_NRF5`` temperature sensor is used, and ``Button 1`` on the board is used to simulate presence detected.

Reviewer: Not sure if the correct highlighting/formatting is used here...
Maybe *TEMP_NRF5*, and *Button 1*, if any formatting at all? :)

.. note::
    This sample must be paired with :ref:`bluetooth_mesh_sensor_client` to show any functionality. It is the Client that gets and uses the sensor data, while the Server only provides it to the Client.

Four Mesh sensor types are used in this sample:

1. :cpp:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the Server according to the publishing period of the Server (see :ref:`bluetooth_mesh_sensor_server_conf_models`).
#. :cpp:var:`bt_mesh_sensor_rel_runtime_in_a_dev_op_temp_range` - Periodically requested by the Client.
#. :cpp:var:`bt_mesh_sensor_presence_detected` - Published when a button is pressed on the Server board.
#. :cpp:var:`bt_mesh_sensor_time_since_presence_detected` - Periodically requested by the Client and published by the Server according to the publishing period of the Server (see :ref:`bluetooth_mesh_sensor_server_conf_models`).

Overview
********

This sample is split into two source files:

* A :file:`main.c` file to handle initialization.
* One additional file for handling Mesh models, :file:`model_handler.c`.

Provisioning
============

> The provisioning is handled by the :ref:`bt_mesh_dk_prov`, and can be done with the `nRF Mesh mobile app`_. It is also possible to configure the models with the app.

Reviewer: ```suggestion
The provisioning is handled by the :ref:`bt_mesh_dk_prov`, and can be done with the `nRF Mesh mobile app`_.
It is also possible to configure the models with the app.
```

Models
======

> The following table shows the Mesh :ref:`bt_mesh_sensor_srv_readme` composition data for this sample:

Reviewer: ```suggestion
The following table shows the Mesh sensor server composition data for this sample:
```

   +---------------+
   |  Element 1    |
   +===============+
   | Config Server |
   +---------------+
   | Health Server |
   +---------------+
   | Sensor Server |
   +---------------+

The models are used for the following purposes:

* Config Server allows configurator devices to configure the node remotely.
* Health Server provides ``attention`` callbacks that are used during provisioning to call your attention to the device.
  These callbacks trigger blinking of the LEDs.

> * :ref:`bt_mesh_sensor_srv_readme` provides sensor data to one or more :ref:`bt_mesh_sensor_cli_readme`.

Reviewer: ```suggestion
* Mesh sensor server provides sensor data to one or more :ref:`sensor client(s) <bt_mesh_sensor_cli_readme>`.
```

#. Select the :guilabel:`Mesh Sensor Server` device to connect to it.
#. Tap :guilabel:`Identify` and then :guilabel:`Provision` to provision the device.
#. When prompted, select an OOB method and follow the instructions in the app.

Once the provisioning is complete, the app returns to the Network screen.

.. _bluetooth_mesh_sensor_server_conf_models:

Configuring models
------------------

Complete the following steps in the nRF Mesh app to configure models:

1. On the Network screen, tap the :guilabel:`Mesh Sensor Server` node.
   Basic information about the Mesh node and its configuration is displayed.
#. In the Mesh node view, expand the element.
   It contains the list of models in the first and only element of the node.
#. Tap :guilabel:`Sensor Server` to see the model's configuration.
#. Bind the model to application keys to make it open for communication:

   1. Tap :guilabel:`BIND KEY` at the top of the screen.
   #. Select :guilabel:`Application Key 1` from the list.

#. Set the publishing parameters:

   1. Tap :guilabel:`SET PUBLICATION`.
   #. Tap :guilabel:`Publish Address`.
   #. Choose Groups from the dropdown menu.

>    #. Choose an existing group or create a new. *IMPORTANT: The Sensor Client must subscribe to the same grop*.

Reviewer 2: Could either publish directly to the client here, or publish to 0xffff to avoid the subscription requirement

Author: Landed on keeping using group address after discussion with Trond

Models
======

The following table shows the Mesh sensor server composition data for this sample:

   +---------------+
   |  Element 1    |
   +===============+
   | Config Server |
   +---------------+
   | Health Server |
   +---------------+
   | Sensor Server |
   +---------------+

The models are used for the following purposes:

* Config Server allows configurator devices to configure the node remotely.
* Health Server provides ``attention`` callbacks that are used during provisioning to call your attention to the device.
  These callbacks trigger blinking of the LEDs.
* Mesh sensor server provides sensor data to one or more :ref:`Mesh sensor client(s) <bt_mesh_sensor_cli_readme>`.

The model handling is implemented in :file:`src/model_handler.c`, which uses the ``TEMP_NRF5`` temperature sensor, and :ref:`dk_buttons_and_leds_readme` to detect button presses.

Requirements
************

* One of the following development kits:

  * |nRF52840DK|
  * |nRF52DK|

* Smartphone with Nordic Semiconductor's nRF Mesh mobile app installed in one of the following versions:

  * `nRF Mesh mobile app for Android`_
  * `nRF Mesh mobile app for iOS`_

> * :ref:`bluetooth_mesh_sensor_client` sample application programmed on a separate device and configured according to the Mesh :ref:`bt_mesh_sensor_cli_readme` sample's :ref:`bluetooth_mesh_sensor_client_testing` guide.

Reviewer: ```suggestion
* :ref:`bluetooth_mesh_sensor_client` sample application programmed on a separate device and configured according to the Mesh sensor client's :ref:`testing guide <bluetooth_mesh_sensor_client_testing>`.
```

User interface
**************

Buttons:
   Can be used to input the OOB authentication value during provisioning.
   All buttons have the same functionality during this procedure.

   In addition, button 1 is used to simulate presence detected.

Building and running
********************

.. |sample path| replace:: :file:`samples/bluetooth/mesh/sensor_server`

.. include:: /includes/build_and_run.txt

.. _bluetooth_mesh_sensor_server_testing:

Testing
=======

> .. important::
>    The :ref:`bt_mesh_sensor_srv_readme` sample cannot demonstrate any functionality on its own, and needs a device with the :ref:`bluetooth_mesh_sensor_client` sample running in the same Mesh network.
>    Before testing Mesh :ref:`bt_mesh_sensor_srv_readme`, go through the Mesh :ref:`bt_mesh_sensor_cli_readme`'s :ref:`bluetooth_mesh_sensor_client_testing` guide with a different board.

Reviewer: ```suggestion
.. note::
   The Mesh sensor server sample cannot demonstrate any functionality on its own, and needs a device with the :ref:`bluetooth_mesh_sensor_client` sample running in the same Mesh network.
   Before testing the Mesh sensor server, go through the Mesh sensor client's :ref:`testing guide <bluetooth_mesh_sensor_client_testing>` with a different board.
```

> After programming the sample to your board, you can test the device by provisioning and configuring it for communication trough the Mesh models.
> This can be done using a smartphone with Nordic Semiconductor's nRF Mesh app installed.

Reviewer: ```suggestion
After programming the sample to your board, you can test it by using a smartphone with Nordic Semiconductor’s nRF Mesh app installed.
Testing consists of provisioning the device and configuring it for communication with the mesh models.
```

Provisioning the device
-----------------------

> The provisioning assigns an address range to the device, and adds it to the Mesh network.

Reviewer: ```suggestion
The provisioning assigns an address range to the device, and adds it to the mesh network.
```
I know it can be confusing regarding Mesh vs mesh. Sometimes we might be in doubt ourselves LOL
Mesh in general terms like here (mesh network) is not capitalized. Capital letter only used when part of a name/term, like (Bluetooth) Mesh sensor server or nRF Mesh app.

Complete the following steps in the nRF Mesh app:

> 1. Tap :guilabel:`Add node` to start scanning for unprovisioned Mesh devices.

Reviewer: ```suggestion
1. Tap :guilabel:`Add node` to start scanning for unprovisioned mesh devices.
```

#. Select the :guilabel:`Mesh Sensor Server` device to connect to it.
#. Tap :guilabel:`Identify` and then :guilabel:`Provision` to provision the device.
#. When prompted, select an OOB method and follow the instructions in the app.

Once the provisioning is complete, the app returns to the Network screen.

.. _bluetooth_mesh_sensor_server_conf_models:

Configuring models
------------------

Complete the following steps in the nRF Mesh app to configure models:

1. On the Network screen, tap the :guilabel:`Mesh Sensor Server` node.

>    Basic information about the Mesh node and its configuration is displayed.

Reviewer: ```suggestion
   Basic information about the mesh node and its configuration is displayed.
```

#. In the Mesh node view, expand the element.
   It contains the list of models in the first and only element of the node.
#. Tap :guilabel:`Sensor Server` to see the model's configuration.
#. Bind the model to application keys to make it open for communication:

   1. Tap :guilabel:`BIND KEY` at the top of the screen.
   #. Select :guilabel:`Application Key 1` from the list.

#. Set the publishing parameters:

   1. Tap :guilabel:`SET PUBLICATION`.
   #. Tap :guilabel:`Publish Address`.

>    #. Choose Groups from the dropdown menu.
>    #. Choose an existing group or create a new.
>          The Sensor Client must subscribe to the same grop.

Reviewer: ```suggestion
   #. Select Groups from the drop-down menu.
   #. Select an existing group or create a new.
       The sensor client must subscribe to the same group.
```

   #. Tap :guilabel:`OK`.

>    #. Choose a publishing period by using the :guilabel:`Interval` slider.

Reviewer: ```suggestion
   #. Select a publishing period by using the :guilabel:`Interval` slider.
```

>    #. Retransmit Count should be 0 to avoid duplicate logging in the :ref:`bt_mesh_sensor_cli_readme` UART terminal.
>    #. Press the :guilabel:`check button` at the bottom right corner.

Reviewer: ```suggestion
   #. Retransmit Count should be 0 to avoid duplicate logging in the :ref:`sensor client's <bt_mesh_sensor_cli_readme>` UART terminal.
   #. Tap the :guilabel:`check button` at the bottom right corner.
```

#. Set subscription parameters:

   1. Tap :guilabel:`SUBSCRIBE`.

>    #. Choose an existing group or create a new.
>          The Sensor Client must publish to the same grop.

Reviewer: ```suggestion
   #. Select an existing group or create a new.
       The sensor client must publish to the same group.
```

   #. Tap :guilabel:`OK`.

#. Tap the :guilabel:`back arrow` at the top left corner.

.. note::

>     If the server is going to be configured by a client, an application key must be bound to the sensor setup server as well.
>     This functionality must also be programmed in the :ref:`bt_mesh_sensor_cli_readme` device.

Reviewer: ```suggestion
    To enable server configuration by a client, an application key must be bound to the sensor setup server.
    This functionality must also be programmed in the :ref:`bt_mesh_sensor_cli_readme` device.
```

.. _bluetooth_mesh_sensor_server:

Bluetooth: Mesh sensor server
###############################

The Bluetooth Mesh sensor server sample demonstrates how to set up a basic Mesh sensor server model application that provides sensor data to one :ref:`Mesh sensor client <bt_mesh_sensor_cli_readme>` model.

.. note::
    This sample must be paired with :ref:`bluetooth_mesh_sensor_client` to show any functionality. The server provides the sensor data used by the client.

Overview
********

This sample is split into the following source files:

* A :file:`main.c` file to handle initialization.
* One additional file for handling Mesh models, :file:`model_handler.c`.

> The following Mesh sensor types are used in this sample:

Reviewer 3: Could you provide a line of explanation after this list or in line 7 (which would be better, I think) about why these various sensor types are used in this sample?

Author: For example "Four different sensor types are used to showcase different ways for the server to publish data. In addition, usage of both single-channel sensors types and series sensor types are shown (reference: http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bluetooth/mesh/sensor.html?#bt-mesh-sensor-types-channels)."

The models are used for the following purposes:

* Config Server allows configurator devices to configure the node remotely.
* Health Server provides ``attention`` callbacks that are used during provisioning to call your attention to the device.
  These callbacks trigger blinking of the LEDs.
* Mesh sensor server provides sensor data to one or more :ref:`Mesh sensor client(s) <bt_mesh_sensor_cli_readme>`.

The model handling is implemented in :file:`src/model_handler.c`, which uses the ``TEMP_NRF5`` temperature sensor, and :ref:`dk_buttons_and_leds_readme` to detect button presses.

Requirements
************

* One of the following development kits:

  * |nRF52840DK|
  * |nRF52DK|

* Smartphone with Nordic Semiconductor's nRF Mesh mobile app installed in one of the following versions:

  * `nRF Mesh mobile app for Android`_
  * `nRF Mesh mobile app for iOS`_

* :ref:`bluetooth_mesh_sensor_client` sample application programmed on a separate device and configured according to the Mesh sensor client's :ref:`testing guide <bluetooth_mesh_sensor_client_testing>`.

User interface
**************

Buttons:
   Can be used to input the OOB authentication value during provisioning.

>    All buttons have the same functionality during this procedure.

Reviewer 3: Hm, isn't Button 1 an exception here?

Author: All buttons have the same functionality during the provisioning procedure

Button 1:

>    Simulate presence detected.

Reviewer 3: Can you rephrase slightly, so that the effect of pressing the button is more pronounced? For example, "Simulate detecting presence to allow ..."?

Author: Button 1 simulates presence detected after the provisioning procedure is finished. Maybe change the 
heading(s)?

Once the provisioning is complete, the app returns to the Network screen.

.. _bluetooth_mesh_sensor_server_conf_models:

Configuring models
------------------

Complete the following steps in the nRF Mesh app to configure models:

1. On the Network screen, tap the :guilabel:`Mesh Sensor Server` node.
   Basic information about the mesh node and its configuration is displayed.
#. In the Mesh node view, expand the element.
   It contains the list of models in the first and only element of the node.
#. Tap :guilabel:`Sensor Server` to see the model's configuration.
#. Bind the model to application keys to make it open for communication:

   a. Tap :guilabel:`BIND KEY` at the top of the screen.
   #. Select :guilabel:`Application Key 1` from the list.

#. Set the publishing parameters:

   a. Tap :guilabel:`SET PUBLICATION`.
   #. Tap :guilabel:`Publish Address`.
   #. Select :guilabel:`Groups` from the drop-down menu.
   #. Select an existing group or create a new one.

      .. note::
            The sensor client must subscribe to the same group.

   #. Tap :guilabel:`OK`.
   #. Select a publishing period by using the :guilabel:`Interval` slider.

>    #. Set the Retransmit Count to zero (:guilabel:`Disabled`) to avoid duplicate logging in the :ref:`sensor client's <bt_mesh_sensor_cli_readme>` UART terminal.

Reviewer 3: See my comment to the client.

Author: Disabled does match the app

>    #. Tap :guilabel:`Check` at the bottom right corner.

Reviewer 3: See my comment to the client.

Author: The button is there with a "check icon" (like this: https://en.wikipedia.org/wiki/File:Light_green_check.svg)

#. Set subscription parameters:

   a. Tap :guilabel:`SUBSCRIBE`.
   #. Select an existing group or create a new one.

      .. note::
            The sensor client must publish to the same group.

   #. Tap :guilabel:`OK`.

> #. Tap the back arrow at the top left corner.

Reviewer 3: See my comment to the client.

Author: This is a "back button" to get back to the previous menu. If it is pressed twice you get back to the "main menu". Maybe it is better to write something like this:
"The sensor server model is now configured and should be able to send data to the client. Press the back arrow (top left corner) twice to get back to the main screen. "
